website/blog/2024-02-29-StateFlow/index.mdx

```diff
@@ -0,0 +1,98 @@
+We also allow passing in a customized speaker selection function to the `speaker_selection_method` to customize the selection rule for next speaker.
+These transitions can based on current speaker and static checking of the context history (for example, checking if 'Error' is in last message).
+
+
+## Example of a custom_speaker_selection_func
+We present an quick example of how to define a custom speaker selection function for a desired workflow.
+Assume we have a task to retrieve papers related to a topic and create a markdown table for summary. 
+
+The workflow is:
+1. Write Python Code to retrieve papers from arxiv.
+2. Execute the code.
+3. Create the markdown table and terminate the conversation.
+
+To implement this workflow, we create 3 agents: `engineer`, `executor`, and `scientist` for each step.
+```python
+# Define the agents, the code is for illustration purposes and is not executable.
+coder = autogen.AssistantAgent(
+    name="Coder",
+    system_message="""You are the Coder. Write Python Code to retrieve papers from arxiv.."""
+)
+executor = autogen.UserProxyAgent(
+    name="Executor",
+    system_message="Executor. Execute the code written by the Coder and report the result.",
+)
+scientist = autogen.AssistantAgent(
+    name="Scientist",
+    system_message="""You are the Scientist. Please categorize papers after seeing their abstracts printed and create a markdown table with Domain, Title, Authors, Summary and Link. Return 'TERMINATE' in the end.""",
+)
+```
+
+Then we define a custom speaker selection function to control the transition between these agents:
+```python
+def custom_speaker_selection_func(last_speaker: Agent, messages: List[Dict]):
+    """Define a customized speaker selection function.
+    A recommended way is to define a transition for each speaker in the groupchat.
+
+    Returns:
+        Return an `Agent` class or a string from ['auto', 'manual', 'random', 'round_robin'] to select a default method to use.
+    """
+    if len(messages) <= 1:
+        # start with the engineer
+        return engineer
+
+    if last_speaker is engineer:
+        if "```python" in messages[-1]["content"]:
+            # If the last message is a python code block, let the executor to speak
+            return executor
+        else:
+            # Otherwise, let the engineer to continue
+            return engineer
+
+    elif last_speaker is executor:
+        if "exitcode: 1" in messages[-1]["content"]:
+            # If the last message indicates an error, let the engineer to improve the code
+            return engineer
+        else:
+            # Otherwise, let the scientist to speak
+            return scientist
+
+    elif last_speaker is scientist:
+        # if the scientist speaks but doesn't terminate the task, we return to a build-in 'auto' selection method that
+        # asks the groupchat manager to select the next speaker.
+        return "auto"
+
+groupchat = autogen.GroupChat(
```

Comment:
The usage of this example seems unrelated to this `Stateflow` PR? (#1816)?

Reply:
Hello @LittleLittleCloud, sorry for the confusion. #1816 is attempting to introduce a completely new concept and still needs discussion. Currently, it is best to accomplish the idea with groupchat, with customized functions from #1791.